Fix `std.os.windows.GetFinalPathNameByHandle` for os versions before `win10_rs4`

rohlem opened the pull request:
This PR fixes `std.os.windows.GetFinalPathNameByHandle` for older Windows versions, unblocking #7242 (although that might be an ongoing effort, so shouldn't be closed yet I think).

#5993 removed the codepath that called the `kernel32.dll` implementation. According to [some document I found](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/efbfe127-73ad-4140-9967-ec6500e66d5e), the new `.FileNormalizedNameInformation` query value was only added as recently as 2018 (not even 3 years ago), meaning we dropped support for more than just Windows 7.

Originally I just wanted to query the runtime OS version and reintroduce the old codepath, but then I tried the [prophesized `NtQueryObject` function](https://github.com/ziglang/zig/pull/5993#issuecomment-693209192), and that ended up working too. Note though that it's using an undocumented value (`ObjectNameInformation = 1` appears in `ntifs.h`, but not the documentation) and a `struct` that is named appropriately but also documented somewhere else. And yet, it still works.

_If_ we trust this new implementation, I would be inclined to maybe delete the other codepath (that needs two `ntdll` calls instead of just one) in favor of this one. (And maybe move the logic for finding DOS volume names into its own function.)
But now that I've worked on this for a day, I thought I would ask for some feedback, to make sure I'm not overlooking anything crucial.

Tested (see new tests in std/os/windows.zig) on my Windows 7 and an up-to-date Windows 10 machine. Things I tried with this implementation:
- NTFS junctions are resolved
- Symbolic links are resolved
- Hard links remain separate paths
- Returned paths > 500 characters work without issue

Also let me know if I should squish/reorganize commits in any way, I thought for documentation/discussion (and maybe unexpected troubleshooting) the commits with the kernel32 path were worth keeping for now.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
-    Unexpected,
-};
+pub fn QueryObjectName(
+    Handle: HANDLE,
```
Lowercase parameter name.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+pub fn QueryObjectName(
+    Handle: HANDLE,
+    out_buffer: []u16,
+    ReturnLength: ?PULONG, //returned in bytes
```
Lowercase parameter name.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    var info = @ptrCast(*OBJECT_NAME_INFORMATION, aligned_raw_buffer);
+    info.* = undefined; //all of the fields we read will be overwritten, and the result string is placed in the same buffer after the struct
+    const out_buffer_length = std.math.cast(ULONG, aligned_raw_buffer.len) catch |e| std.math.maxInt(u32); //we'll just use as much of it as we can
+    const rc = ntdll.NtQueryObject(Handle, .ObjectNameInformation, out_buffer.ptr, out_buffer_length, ReturnLength); //buffer size is specified in u16...
```
The code is extremely hairy, it takes an output buffer `out_buffer` where the pathname is stored _and_ uses it to allocate a `OBJECT_NAME_INFORMATION` too.

I'd simply allocate a buffer that's hopefully large enough (path_max plus the `OBJECT_NAME_INFORMATION` overhead) and copy the result over.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    info.* = undefined; //all of the fields we read will be overwritten, and the result string is placed in the same buffer after the struct
+    const out_buffer_length = std.math.cast(ULONG, aligned_raw_buffer.len) catch |e| std.math.maxInt(u32); //we'll just use as much of it as we can
+    const rc = ntdll.NtQueryObject(Handle, .ObjectNameInformation, out_buffer.ptr, out_buffer_length, ReturnLength); //buffer size is specified in u16...
+    switch(rc){
```
The missing spaces hint you're not using `zig fmt`, please run it on all the files this PR is touching.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    if (comptime builtin.os.tag != .windows)
+        return;
+    
+    const file: std.fs.File = try std.fs.openSelfExe(.{}); //any file will do; canonicalization works on NTFS junctions and symlinks, hardlinks remain separate paths.
```
Redundant types are redundant.
And please keep the comments in a line above the one they refer to, not everyone has a huge monitor.

```suggestion
    //any file will do; canonicalization works on NTFS junctions and symlinks, hardlinks remain separate paths.
    const file = try std.fs.openSelfExe(.{}); 
```

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    else
+        error {
+            AccessDenied,
+            BufferTooSmall,
```
`NameTooLong` and `BufferTooSmall` ?

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+            AccessDenied,
+            BufferTooSmall,
+            InvalidHandle,
+            Unnamed,
```
Unnamed ?

rohlem replied:
That possibility is mentioned [here](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-obquerynamestring) (`if the given object is unnamed`), which I found used in the implementation of ReactOS's `NtQueryObject`. I don't know when it would be triggered though, might be for other "objects" than files? I'll change that to unreachable.

kubkon replied:
Hmm, could we not coerce `Unnamed` as simply `Unexpected`? Or put it another way, what value will conveying that the object is unnamed give to the user?

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1032,54 @@ pub fn GetFinalPathNameByHandle(
-        .Nt => {
+        //otherwise we need to parse the string for volume path for the .Dos logic below to work
+        const expected_prefix = std.unicode.utf8ToUtf16LeStringLiteral("\\Device\\");
+        std.debug.assert(std.mem.eql(u16, expected_prefix, final_path[0..expected_prefix.len]));
```
This is elided for non-debug builds, is this intentional?

rohlem replied:
Well, it generates `unreachable`. I don't know enough about NT paths to judge whether/where this can happen, and what to do if it does.
I suppose returning an error is safer? And I think `error.Unexpected`, since it's caused by the system and `error.BadPathName` sounds like a caller error?

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1032,54 @@ pub fn GetFinalPathNameByHandle(
+        std.debug.assert(std.mem.eql(u16, expected_prefix, final_path[0..expected_prefix.len]));
+        const index = std.mem.indexOfPos(u16, final_path, expected_prefix.len, &[_]u16{'\\'}) orelse unreachable;
+        volume_name_u16 = final_path[0..index];
+        file_name_u16 = final_path[index..];
```
This gets the `\`, is it intentional?

rohlem replied:
When I tested it seemed to give the same results as the other codepath, so I would say yes. Also the docs say absolute paths start with `\`, while relative paths don't. (`C:test\abc.txt` is a relative path.)

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -1091,6 +1166,43 @@ pub fn GetFinalPathNameByHandle(
+    const S = struct {fn check(file: std.fs.File, buffer: anytype, fmt: std.os.windows.GetFinalPathNameByHandleFormat) !usize {
+        const real = try std.os.windows.GetFinalPathNameByHandle(file.handle, fmt, buffer[0..]);
+        const offset = @divExact((@ptrToInt(real.ptr) - @ptrToInt(@ptrCast([*]u16, buffer))), 2);
+        return real.len + offset;
```
What's this `offset` for? Why do you care about what's before `real.ptr`?

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -1091,6 +1166,43 @@ pub fn GetFinalPathNameByHandle(
+    const dos_length = try S.check(file, buffer[0..], .{.volume_name = .Dos});
+
+    //check with size insufficient by more than 1 character: always errors
+    std.testing.expect(null == S.check(file, buffer[0..nt_length-1], .{.volume_name = .Nt})
```
```suggestion
    std.testing.expectEqual(null, S.check(file, buffer[0..nt_length-1], .{.volume_name = .Nt})
```

Ditto everywhere else, this way if the check fails it'll show you the faulty value.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -1091,6 +1166,43 @@ pub fn GetFinalPathNameByHandle(
+    //check with size insufficient for null terminator: errors only on NtQueryObject path
+    const queryObjectPath = !std.os.windows.runtimeVersionIsAtLeast(std.os.windows.WindowsVersion.win10_rs4);
+    std.testing.expect(queryObjectPath == (null == S.check(file, buffer[0..nt_length], .{.volume_name = .Nt})
+        catch |e| switch(e) {error.NameTooLong => null, else => unreachable}));
```
There's `testing.expectError` for this.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -1601,6 +1713,67 @@ inline fn MAKELANGID(p: c_ushort, s: c_ushort) LANGID {
+
+    switch (std.os.windows.ntdll.RtlGetVersion(&version_info)) {
+        .SUCCESS => {},
+        else => |e| { //only return value STATUS_SUCCESS is documented
```
```suggestion
        else => unreachable
```

This API cannot fail according to the docs.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -1601,6 +1713,67 @@ inline fn MAKELANGID(p: c_ushort, s: c_ushort) LANGID {
+
+/// Returns whether the runtime os version is at least as new as the argument.
+pub fn runtimeVersionIsAtLeast(requested_version: WindowsVersion) bool {
+    return if(targetVersionIsAtLeast(requested_version)) |fully| fully
```
```suggestion
    return if(targetVersionIsAtLeast(requested_version)) orelse
        @enumToInt(detectRuntimeVersion()) >= @enumToInt(requested_version);
```

rohlem commented:
Thanks for the quick review, I might only get to it tomorrow though.

kubkon commented on lib/std/os/windows.zig:
```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+        NameTooLong,
+        Unexpected,
+    }
+    || if((comptime builtin.os.tag != .windows) or (targetVersionIsAtLeast(WindowsVersion.win10_rs4) == true))
```
Given @LemonBoy's comments below, if we collapse a lot of these error values into the ones already existing, is it necessary to comptime check for Windows version, or would it be more readable to slightly enlarge the return error set instead? We already do that in a few places in `std.os` mind where some error values are WASI or Windows-only but they're part of the same return set.

kubkon commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1032,54 @@ pub fn GetFinalPathNameByHandle(
+        const volume_name_info = @ptrCast(*const FILE_NAME_INFORMATION, &volume_buffer[0]);
+        volume_name_u16 = @ptrCast([*]const u16, &volume_name_info.FileName[0])[0 .. @divExact(volume_name_info.FileNameLength, 2)];
+
+        if(fmt.volume_name == .Nt){
```
Much like the codepath below, I think a `switch` statement here with an `unreachable` for `.Dos` would convey the meaning a little better than an `if` statement without an `else`.

rohlem replied:
`.Dos` is not `unreachable`, it falls through to the rest of the function. Or are you suggesting this?
```zig
switch(fmt){
  .Nt => {...},
  .Dos => {}//fallthrough
}
```
?

kubkon replied:
Ah gotcha. It's fine as-is then.

kubkon commented on lib/std/os/windows.zig:
```diff
@@ -1601,6 +1713,67 @@ inline fn MAKELANGID(p: c_ushort, s: c_ushort) LANGID {
+
+/// Returns whether the target os versions are uniformly at least as new as the argument:
+/// true/false if this holds for the entire target range, null if it only holds for some.
+pub fn targetVersionIsAtLeast(requested_version: WindowsVersion) ?bool {
```
Should these not belong in `system.zig`?

kubkon commented on lib/std/os/windows.zig:
```diff
@@ -1601,6 +1713,67 @@ inline fn MAKELANGID(p: c_ushort, s: c_ushort) LANGID {
+
+/// Returns the highest known WindowsVersion deduced from reported runtime information.
+/// Discards information about in-between versions we don't differentiate.
+pub fn detectRuntimeVersion() WindowsVersion {
```
Is this call really necessary? We already call `std.os.windows.ntdll.RtlGetVersion` in `std.zig.system` and in this line here in particular: [std/zig/system.zig#L229](https://github.com/ziglang/zig/blob/master/lib/std/zig/system.zig#L229). Also, for future reference, any bit of code that works out the target/crosstarget version should go in `system.zig`.

kubkon commented on lib/std/zig/system.zig:
```diff
@@ -223,44 +223,9 @@ pub const NativeTargetInfo = struct {
-
-                    os.version_range.windows.max = @intToEnum(Target.Os.WindowsVersion, version);
-                    os.version_range.windows.min = @intToEnum(Target.Os.WindowsVersion, version);
+                    const detected_version = std.os.windows.detectRuntimeVersion();
```
Oh I see now. This function should still be part of the `system.zig` though. I suggest putting it in `lib/std/zig/system/windows.zig` much like we have macOS specific target detection in `lib/std/zig/system/macos.zig`. @andrewrk's idea which I agree with is to have all version detection functionality in solely in the `system` module and nowhere else.

marler8997 commented:
Just wanted to say thank you to @rohlem for showing windows some love, it desperately needs it.

rohlem commented:
Rebased onto master and integrated all review comments, sorry it took so long.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
-            const volume_name_u16 = @ptrCast([*]const u16, &volume_name.FileName[0])[0 .. volume_name.FileNameLength / 2];
+        //otherwise we need to parse the string for volume path for the .Dos logic below to work
+        const expected_prefix = std.unicode.utf8ToUtf16LeStringLiteral("\\Device\\");
+        if (!std.mem.eql(u16, expected_prefix, final_path[0..expected_prefix.len])) {
```
Given this is a precondition we expect the kernel to respect, an assert is fine.

Please use `mem` instead of `std.mem`!

rohlem replied:
Well, I'm not knowledgeable enough about these NT paths to say whether the returned string always starts with `\Device\`, it's just the only scenario I've encountered, but I don't have any network sharing set up f.e. .

`BadPathName` is already returned further down if the DOS volume name we've resolved unexpectedly contains null characters, which also seems like a "not-yet-disproven working thesis", so I think I'll stick to it here for consistency.
But let me know if you'd prefer the assert here, or in both places.

rohlem replied:
Rethinking it yet again, the way we just blindly assume the prefix could lead to nonsensical results in unsafe build mode. An error return makes much more sense to me. But speak up if you feel differently.

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
+        volume_name_u16 = final_path[0..index];
+        file_name_u16 = final_path[index..];
+
+        //fallthrough for fmt.volume_name != .Nt
```
fallthrough? The other case is handled in the other branch.

rohlem replied:
No, the "other branch" (the else at line `1065` of the if at line `1037`) is for the newer windows version... unless I'm missing something?

LemonBoy commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
+        const volume_name_info = @ptrCast(*const FILE_NAME_INFORMATION, &volume_buffer[0]);
+        volume_name_u16 = @ptrCast([*]const u16, &volume_name_info.FileName[0])[0..@divExact(volume_name_info.FileNameLength, 2)];
 
+        if (fmt.volume_name == .Nt) {
```
This is in the `(fmt.volume_name != .Nt)` branch, it'll never be executed.

rohlem replied:
No, this is the `win10_rs4 or newer` branch, the previous code I kept intact.

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    //last argument would return the length required for full_buffer, not exposed here
+    const rc = ntdll.NtQueryObject(handle, .ObjectNameInformation, full_buffer[0..], full_buffer_length, null);
+    return switch (rc) {
+        .SUCCESS => if (@ptrCast(?[*]WCHAR, info.Name.Buffer)) |buffer| blk: {
```
[The documentation for ObQueryNameString](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-obquerynamestring) , which I based this on since it seems to be used under the hood, mentions a `NULL` value under "Remarks" (`If the given object is unnamed [...]`).
`info.MaxLength` is also set to `0` in that case, so I'm happy to change the check to that, but returning an empty string would require checking for that in the consumer code (`GetFinalPathNameByHandle`).
The object's name isn't the empty string in this case, it has no name - I think either returning an optional or error still makes more sense.

kubkon replied:
I'm still on the fence with the idea of having `error.Unexpected`. If something like this happens, shouldn't we simply trigger a nonrecoverable runtime trap instead? Or do we intend for user-space code to be recoverable in those circumstances? If that's the case, anyone's got any ideas for a scenario when that would be the case?

rohlem replied:
"nonrecoverable runtime trap" (so basically `@panic`?) sounds very cut-throat to me.
If the file was supplied by a user, say interactively via `stdin`, maybe you just want to notify the user "this file didn't work, try another / moving it to a different path" (or something more specific if you figure out the underlying problem).
To me this seems like a fitting scenario for an error code: We failed to provide the requested service, but if/how to recover is up to the caller. That's why I made it return `error.Unnamed` before.
If we don't expect this API to be used with unnamed objects, then this scenario is "unexpected". `error.Unexpected` is not inherently reserved for kernel failures, so I think it fits.

But these are just my thoughts, whatever you think is best I'll implement.

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    var full_buffer: [@sizeOf(OBJECT_NAME_INFORMATION) + PATH_MAX_WIDE * 2]u8 align(@alignOf(OBJECT_NAME_INFORMATION)) = undefined;
+    var info = @ptrCast(*OBJECT_NAME_INFORMATION, &full_buffer);
+    //buffer size is specified in bytes
+    const full_buffer_length = @intCast(ULONG, @sizeOf(OBJECT_NAME_INFORMATION) + std.math.min(PATH_MAX_WIDE, (out_buffer.len + 1) * 2));
```
Hmm, you're right, though now that I look at it again, with this exact size specification we could conversely remove the size check below.
But I guess the explicit check below is better in the sense of readability?

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+            if (out_buffer.len < path_length_unterminated) {
+                return error.NameTooLong;
+            }
+            std.mem.copy(WCHAR, out_buffer[0..path_length_unterminated], buffer[0..path_length_unterminated :0]);
```
Huh, good eye. Strange to document that with the data structure, but I guess that means it could apply anywhere it's used...

Mouvedia commented on lib/std/target.zig:
```diff
@@ -109,6 +109,20 @@ pub const Target = struct {
+                17134, //win_rs4
+                17763, //win_rs5
+                18362, //win_19h1
+                19041, //win_20h1
```
You forgot 2 builds: `18363, 19042`

cf https://github.com/ziglang/zig/commit/140c599559b6e048db2ec8476cb2cc08366f6aa8#r44615067

rohlem replied:
I thought so at first too, but those don't impact [the `NTDDI` header constant we're trying to mirror here](https://docs.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt). (Also the values are unchanged from the previous implementation.)

Mouvedia replied:
This page is dated 2020-06-19, `19042` was released the 2020-10-20.
i.e. it's outdated

rohlem replied:
Okay, it seems that you're right, [this repository's own `sdkddkver.h`](https://github.com/ziglang/zig/blob/master/lib/libc/include/any-windows-any/sdkddkver.h#L144-L146) already includes more constants than I found when I was looking for them earlier. Thank you for pointing this out!

Mouvedia replied:
FYI the next one will probably be 20231 (21H1)

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    const file = try std.fs.openSelfExe(.{});
+    defer file.close();
+    //make this large enough for the test runner exe path
+    var out_buffer align(16) = std.mem.zeroes([1 << 10]u16);
```
Right, that is a leftover from earlier.
And sure, if we have the stack space for another PATH_MAX_WIDE that's fine by me.

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
-    const file_name_u16 = @ptrCast([*]const u16, &file_name.FileName[0])[0 .. file_name.FileNameLength / 2];
+    var file_name_u16: []const u16 = undefined;
+    var volume_name_u16: []const u16 = undefined;
+    if ((comptime (targetVersionIsAtLeast(WindowsVersion.win10_rs4) != true)) //need explicit comptime, because error returns affect return type
```
It looks strange, but the value is of optional type `?bool`, not `bool` directly.
So this is supposed to catch both `null` and `false`. I don't think either of `!` and `== false` work for this.

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
+        and !runtimeVersionIsAtLeast(WindowsVersion.win10_rs4))
+    {
+        const final_path = QueryObjectName(hFile, std.mem.bytesAsSlice(u16, path_buffer[0..])) catch |err| return switch (err) {
+            error.InvalidHandle => error.FileNotFound, //close enough?
```
Close to the semantic meaning of "invalid file handle supplied" - assuming that a not found file be linked to an invalid handle.

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
+    var file_name_u16: []const u16 = undefined;
+    var volume_name_u16: []const u16 = undefined;
+    if ((comptime (targetVersionIsAtLeast(WindowsVersion.win10_rs4) != true)) //need explicit comptime, because error returns affect return type
+        and !runtimeVersionIsAtLeast(WindowsVersion.win10_rs4))
```
I was actually wondering whether we should cache that version in a global (that is, module-scope) variable - not just for this code path, but for all of `std.os.windows`, but it seemed like too big of a change to just throw in unsupervised.
So would an atomic version number `u32` in `std.zig.system.windows` be fitting, initialized once and then used instead of the `RtlGetVersion` call?
Or is there some benefit to having this particular flag be in its own variable directly?

kubkon replied:
Personally, I'm not sure I'm a big fan of having a global/thread-local state lying around, but on the other hand, calling `RtlGetVersion` every single time we execute `GetFinalPathNameByHandle` is a performance hog indeed. @andrewrk what are your thoughts on this?

rohlem replied:
@kubkon Right, that's why I originally refrained.
Maybe this is also a good place to mention that (A) we can always incrementally improve this in subsequent commits, and (B) `NtQueryObject` works for ALL windows versions (for now).
So if just removed the other path (or are satisfied with only the exhaustive `comptime` check), the runtime check instantly becomes superfluous.

daurnimator replied:
`RtlGetVersion` should be pretty cheap: its a read from the PEB.
We could read from the PEB ourselves too, but there shouldn't be much of a difference.

rohlem commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
+                return error.NameTooLong;
+            }
+            std.mem.copy(u16, out_buffer[0..], final_path[0..]);
+            return final_path; //we can directly return the slice we received
```
Ah, right, meant out_buffer of course, thanks.

kubkon commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    //make this large enough for the test runner exe path
+    var out_buffer align(16) = std.mem.zeroes([1 << 10]u16);
+
+    var result_path = try QueryObjectName(file.handle, out_buffer[0..]);
```
Huh, that's brilliant! Never thought of that 😅

andrewrk commented on lib/std/os/windows/bits.zig:
```diff
@@ -65,6 +65,7 @@ pub const SIZE_T = usize;
 pub const TCHAR = if (UNICODE) WCHAR else u8;
 pub const UINT = c_uint;
 pub const ULONG_PTR = usize;
+pub const PULONG = *ULONG;
```
Prefer to use `*ULONG` directly rather than `PULONG`

marler8997 replied:
Yeah, Zig has more "specific" ways of identifying pointers `*ULONG` vs `[*]ULONG`.  `PULONG` is more ambiguous.

andrewrk commented on lib/std/os/windows.zig:
```diff
@@ -28,6 +28,9 @@ pub const gdi32 = @import("windows/gdi32.zig");
 
 pub usingnamespace @import("windows/bits.zig");
 
+//version detection
+usingnamespace std.zig.system.windows;
```
Prefer to avoid `usingnamespace` unless it's to set up the public declarations (as above). Here we should give it a name and refer to the functions therein via field access syntax.

andrewrk commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    const full_buffer_length = @intCast(ULONG, @sizeOf(OBJECT_NAME_INFORMATION) + std.math.min(PATH_MAX_WIDE, (out_buffer.len + 1) * 2));
+    //last argument would return the length required for full_buffer, not exposed here
+    const rc = ntdll.NtQueryObject(handle, .ObjectNameInformation, full_buffer[0..], full_buffer_length, null);
+    return switch (rc) {
```
When returning errors, prefer to put `return` on multiple lines. This will make the error return trace point nicely to the error code that triggered the crash, rather than pointing to `return switch (rc) {`.

andrewrk commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    handle: HANDLE,
+    out_buffer: []u16,
+) ![]u16 {
+    var full_buffer: [@sizeOf(OBJECT_NAME_INFORMATION) + PATH_MAX_WIDE * 2]u8 align(@alignOf(OBJECT_NAME_INFORMATION)) = undefined;
```
Why 2 buffers with a memcpy? We could pass out_buffer directly to NtQueryObject right?

rohlem replied:
Yes, that's what the code did originally, which was commented on [here](https://github.com/ziglang/zig/pull/7379#discussion_r540323038), so I changed it.
I can change it back if you want - I don't think there was a particular reason besides readability - but I don't like stepping on anyone's toes, so I'll put that in as the last commit and then we can pick whichever turns out looking nicer.

andrewrk commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+        return;
+
+    //any file will do; canonicalization works on NTFS junctions and symlinks, hardlinks remain separate paths.
+    const file = try std.fs.openSelfExe(.{});
```
Better not to entangle this test with the functionality of openSelfExe. I suggest to use `std.testing.tmpDir`. You can see example usage in `std/fs/test.zig`.

andrewrk commented on lib/std/os/windows.zig:
```diff
@@ -952,7 +955,53 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+    const file = try std.fs.openSelfExe(.{});
+    defer file.close();
+    //make this large enough for the test runner exe path
+    var out_buffer align(16) = std.mem.zeroes([1 << 10]u16);
```
what is the purpose of calling mem.zeroes here?

rohlem replied:
I think that was a "sane default" during early testing that I forgot to remove. The code shouldn't rely on the value, I'll change it to undefined.

andrewrk commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
+        };
 
-    const volume_name = @ptrCast(*const FILE_NAME_INFORMATION, &volume_buffer[0]);
+        if (fmt.volume_name == .Nt) {
```
Prefer to `switch` on enums rather than `if`

Side note @kubkon - why is `GetFinalPathNameByHandleFormat ` a struct containing an enum, and not just an enum?

andrewrk commented on lib/std/os/windows.zig:
```diff
@@ -980,31 +1029,64 @@ pub fn GetFinalPathNameByHandle(
+        //fallthrough for fmt.volume_name != .Nt
+    }
+
+    switch (fmt.volume_name) {
+        .Nt => unreachable, //handled above
```
This construct can be replaced with `assert(fmt.volume_name == .Nt);`

andrewrk commented on lib/std/zig/system/windows.zig:
```diff
@@ -0,0 +1,61 @@
+
+/// Returns whether the target os versions are uniformly at least as new as the argument:
+/// true/false if this holds for the entire target range, null if it only holds for some.
+pub fn targetVersionIsAtLeast(requested_version: WindowsVersion) ?bool {
+    const requested = @enumToInt(requested_version);
+    const version_range = std.builtin.os.version_range.windows;
+    const target_min = @enumToInt(version_range.min);
+    const target_max = @enumToInt(version_range.max);
+    return if (target_max < requested) false else if (target_min >= requested) true else null;
+}
```
This looks like duplicated code from `std.Target.Os.isAtLeast`

andrewrk reviewed (changes requested):
Thank you for taking on this project, @rohlem. This is good stuff. Needs a bit of cleanup, but the important parts are in place.

> _If_ we trust this new implementation, I would be inclined to maybe delete the other codepath (that needs two `ntdll` calls instead of just one) in favor of this one. (And maybe move the logic for finding DOS volume names into its own function.)
> But now that I've worked on this for a day, I thought I would ask for some feedback, to make sure I'm not overlooking anything crucial.

If it works empirically (which it sounds like it does) then I would be in favor of deleting the other codepath and deleting the version checking.

kubkon commented:
I should probably explain what the motivation was for using `FileNormalizedNameInformation` instead of simply resorting to `NtQueryObject`. When I originally had a stab at this, I did a series of tests and comparisons to see how Windows user-level APIs actually do it using the Process Monitor to track the syscalls. It turned out that on my version of Win10, `GetFinalPathNameByHandleW` was calling `NtQueryInformationFile` with `FileNormalizedNameInformation` file information class. My guess was that this way we wouldn't have to worry about the maximum path name limitation, i.e., that the newer API would handle this for us. Having said that, it is true that ReactOS uses exclusively `NtQueryObject` to implement their version of `GetFinalPathNameByHandleW`. Also, I don't like the fact we have to check Windows version every time we call `GetFinalPathNameByHandle`. On the other hand, @daurnimator reassures us that it should be a pretty cheap call, so maybe this won't be a huge performance hog, but I'm not an expert here by any means. Anyhow, I'm happy if we remove the original alternate path and use `NtQueryObject` exclusively. If this poses some problems to do with path length limits down the line, we can always bring it back or come up with a better solution.

rohlem commented:
Addressed all review comments, besides the one suggested `error -> assert` I commented on - that looks safer as an error to me (not blindly continuing in non-debug build modes).
The last commit removes all intermediate buffers, as @andrewrk suggested (... iiuc), to make the code more readable I added a helper function `std.mem.alignInSlice` to do the necessary in-place-alignment.

LemonBoy commented:
After 83 comments it looks good now!

> The last commit removes all intermediate buffers, as @andrewrk suggested (... iiuc), to make the code more readable I added a helper function std.mem.alignInSlice to do the necessary in-place-alignment.

The helper function makes the code much easier to read and review, thanks.

andrewrk commented:
Landed in 56c03881ebd8617e6506bfad01bf9cfdd4d3df7e
